<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import warnings

## Удаление
удаление ячейки с данными соединения!!!

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [3]:
query_3_1 = f'''
    select count(*)
        from public.vacancies
             '''
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\2542217645.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

##### результат запроса
49197 - количество вакансий

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [4]:
query_3_2 = f'''
    select count(*)
        id
            from public.employers
             '''
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\4087076142.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


id
0  23501

##### результат запроса
23501 - количество работодателей

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [5]:
query_3_3 = f'''
    select count(*)
            from public.areas
             '''
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\212167392.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

##### результат запроса
1362 - количество регионов

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [6]:
query_3_4 = f'''
    select count(*)
            from public.industries
             '''
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\1066549252.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

##### результат запроса
294 - количество сфер деятельности

***

##### выводы по предварительному анализу данных
Первоначальный взгляд на таблицы базы данных project_sql показал нам, что мы имеем дело с:

49197 вакансий,
23501 работодателей,
Работа может предлагаться в 1362 городах, именно столько городов есть в БД,
Работа может быть предоставлена в 294 различных сферах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [7]:
query_4_1 = f'''
    select 
      a.name as area
      ,count(*) as cnt 
    from public.areas as a
    inner join public.vacancies as v on a.id = v.area_id
    group by a.name
    order by cnt desc
    limit 5
      '''
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\2852514614.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

##### результат запроса
По данным вывода мы видим следующее:
В Москве вакансий 5333 (и это максимальное значение),
В Санкт-Петербурге 2851,
В Минске 2112,
В Новосибирске 2006,
В Алмате 1892.

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [8]:
query_4_2 = f'''
    select 
        count(*) as cnt
    from public.vacancies as v
    where v.salary_from is not null 
        or v.salary_to is not null  
      '''
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\3051259067.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

##### результат запроса
У 24073 вакансий хотя бы одно из полей с зарплатой заполнено

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [9]:
query_4_3 = f'''
    SELECT 
        avg(salary_from) as avg_salary_from
        ,avg(salary_to) as avg_salary_to
    FROM public.vacancies as v
      '''
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\2695573716.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0     71064.657901  110536.741923

##### результат запроса
Средняя предлагаемая зарплата находится в диапазоне 71065 - 110537 рублей

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [10]:
query_4_4 = f'''
    select 
        v.schedule,
        v.employment,
        count(*) as vacancies_count
    from public.vacancies as v
    group by v.schedule, v.employment
    order by vacancies_count desc
      '''
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\3724907661.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancies_count
0        Полный день     Полная занятость            35367
1   Удаленная работа     Полная занятость             7802
2      Гибкий график     Полная занятость             1593
3   Удаленная работа  Частичная занятость             1312
4     Сменный график     Полная занятость              940
5        Полный день           Стажировка              569
6     Вахтовый метод     Полная занятость              367
7        Полный день  Частичная занятость              347
8      Гибкий график  Частичная занятость              312
9        Полный день     Проектная работа              141
10  Удаленная работа     Проектная работа              133
11     Гибкий график           Стажировка              116
12    Сменный график  Частичная занятость              101
13  Удаленная работа           Стажировка               64
14     Гибкий график     Проектная работа               18
15    Сменный график           Стажировка               12
16    Вахтовый метод     Проектная работа                2
17    Сменный график     Проектная работа                1

##### результат запроса
Из представленой таблицы на полный рабочий день и полную занятость размещено 35367 вакансий,
далее следует удаленная работа с полной занятостью - 7802 вакансии.
Всего 1 вакансия для сменного графика и проектную работу.

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [11]:
query_4_5 = f'''
    select 
        v.experience
        ,count(*) as vacancies_count
    from public.vacancies as v
    group by v.experience
    order by vacancies_count asc
      '''
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\2992647146.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancies_count
0         Более 6 лет             1337
1           Нет опыта             7197
2       От 3 до 6 лет            14511
3  От 1 года до 3 лет            26152

##### результат запроса
Из представленного графика большинство вакансий с требованием опыта работы от 1 года до 3 лет - 26152,
минимальное количество - 1337 с требованием опыта работы более 6 лет.

***

##### выводы по детальному анализу вакансий

Наибольшее количество предлагаемых ваканский относится к крупным городам.
Максимальное количество вакансий в Москве - 5333, далее следуют Санкт-Петербург - 2851, Минск - 2112, Новосибирск - 2006, Алматы - 1892.
Средняя предлагаемая зарплата находится в диапазоне 71065 - 110537 рублей.
Большая часть вакансий имеет требования по опыту работы от 1 года до 3 лет - 26152 вакансии,
затем с опытом работы от 3 до 6 лет - 14511 вакансий, без опыта - 7197 вакансий и с опытом работы более 6 лет - 1337 вакансий.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [12]:
query_5_1 = f'''
    (select 
        e.name
        ,count(*) as vacancies_count
    from public.vacancies as v
    inner join public.employers as e on e.id = v.employer_id
    group by e.name
    order by vacancies_count desc
    limit 1)
    
    union
    
    (select 
        e.name
        ,count(*) as vacancies_count
    from public.vacancies as v
    inner join public.employers as e on e.id = v.employer_id
    group by e.name
    order by vacancies_count desc
    limit 1
    offset 4)
    order by name desc
      '''
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\144798383.py:24: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  vacancies_count
0         Яндекс             1933
1  Газпром нефть              331

##### результат запроса
На первом месте по количеству вакансий является Яндекс - 1933 вакансии,
на пятом Газпром нефть - 331 вакансия.

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [13]:
query_5_2 = f'''
    select
        a.name 
        ,count(distinct e.id) as employers_count
        ,count(distinct v.id) as vacancies_count
    from public.areas as a
    left join public.employers as e on e.area = a.id
    left join public.vacancies as v on v.area_id = a.id
    where v.id is null
    group by a.name
    order by employers_count desc
    limit 1
      '''
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\2975089403.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employers_count  vacancies_count
0  Россия              410                0

##### результат запроса
Довольно странный вывод, Россия указывается в качестве некоторого района.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [14]:
query_5_3 = f'''
    select
        e.name 
        ,count(distinct v.area_id) as areas_count
    from public.employers as e
    left join public.vacancies as v on v.employer_id = e.id
    group by e.name
    order by areas_count desc
      '''
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\571776059.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  areas_count
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
23170                  СДЕЛКА            0
23171            Alandr Group            0
23172                     СДК            0
23173      Lemon Land Lombard            0
23174                  Махина            0

[23175 rows x 2 columns]

##### результат запроса
Из представленных данных видно, что на первом месте по количеству регионов в которых компании разместили вакансии идет Яндекс - 181 регион, затем Ростелеком - 152 региона, затем Спецремонт - 116 регионов. А вот следующие данные настораживают, некий Поляков Денис Иванович разместил вакансии в 88 регионах, на мой взгляд это похоже на индивидуального предпринимателя, но слишком странно для ИП иметь настолько обширную сеть.

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [15]:
query_5_4 = f'''
    select
      count(e.id) as cnt
    from public.employers as e
    left join public.employers_industries as ei on ei.employer_id = e.id
    where ei.employer_id is null
      '''
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\1336735190.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


cnt
0  8419

##### результат запроса
Из вывода - 8419 работодателей не указали сферу деятельности.

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [16]:
query_5_5 = f'''
    select
      e.name
      ,count(distinct ei.industry_id) as cnt
    from public.employers as e
    left join public.employers_industries as ei on ei.employer_id = e.id
    where ei.employer_id is not null
    group by e.name
    having count(distinct ei.industry_id) = 4
    order by e.name asc
    limit 1
    offset 2
      '''
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\3031831201.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  cnt
0  2ГИС    4

##### результат запроса
Согласно условиям задания, компания находящайся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности - 2ГИС(ДубльГИС).

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [17]:
query_5_6 = f'''
    select
      count(e.name) as cnt
    from public.employers as e
    left join public.employers_industries as ei on ei.employer_id = e.id
    inner join public.industries as i on i.id = ei.industry_id
    where i.name = 'Разработка программного обеспечения'
      '''
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\3793326236.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


cnt
0  3553

##### результат запроса
У 3553 работодателей сфера деятельности указана как 'Разработка программного обеспечения'.

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [18]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
cities = list(pd.read_html(url)[1]['Город'])
cities[13] = cities[13][:-3]
company = 'Яндекс'

load_sql = f'''from public.employers as e
    left join public.vacancies as v on v.employer_id = e.id
    inner join public.areas as a on v.area_id = a.id 
    where 
          e.name = '{company}'
      and v.id is not null
      and a.name in ({','.join(f'{chr(0x27)}{c}{chr(0x27)}' for c in cities)})
    '''
# Долго думал как получить список городов по условию, в Slack подсказали как решить проблему с помощью преобразования SQL-запроса в переменную

In [19]:
query_5_7 = f'''
    (select
      a.name
      ,count(v.id) as cnt
    {load_sql}
    group by 1
    order by 2 asc)
    
    union
    
    (select
      'Total'
      ,count(v.id) as cnt
        {load_sql}
    )
    order by 2
      '''
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\3746532297.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

##### результат запроса
Из данных: Количество регионов-миллионников в которых компания Яндекс разместила вакансии: на первом месте Москва - 54 вакансии, затем Санкт-Петербург - 42, Екатеринбург - 39, Нижний Новгород - 36, Новосибирск - 35, Воронеж - 32, Краснодар - 30, Самара - 26, Уфа - 26, Ростов-на-Дону - 25, Казань - 25, Пермь - 25, Волгоград - 24, Красноярск - 23, Челябинск - 22, Омск - 21. Общее значение на эти города - 485 вакансий.

***

##### выводы по анализу работодателей

В данной базе присутствуют как крупные работодатели имеющие различные вакансии в разных регионах, так и небольшие.
8419 работодателей не указали сферу деятельности, что проблематично при работе с базой, так как направление компании для некоторых соискателей играет свою роль.
Некий Поляков Денис Иванович разместил вакансии в 88 регионах, на мой взгляд это похоже на индивидуального предпринимателя, но слишком странно для ИП иметь настолько обширную сеть.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [20]:
query_6_1 = f'''
    select
      count(distinct v.id) as cnt
    from public.vacancies as v
    where v.name ~~* any (array[
      '%data%'
      ,'%данн%'
    ])
      '''
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\3035133598.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

##### результат запроса
1771 вакансия имеет отношение к данным, учитывая условие, что в её названии имеются слова 'data' или 'данн'.

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [21]:
query_6_2 = f'''
    select
    count(v.name)
    from vacancies v
    where (lower(v.name) like '%data%scientist%'
    or lower(v.name) like '%data%science%'
    or lower(v.name) like '%исследователь%данных%'
    or (lower(v.name) ilike '%ML%'
    and lower(v.name) not ilike '%HTML%')
    or lower(v.name) like '%machine%learning%'
    or lower(v.name) like '%машинн%обучен%')
    and (lower(v.experience) like '%нет опыта%' 
    or lower(v.employment) like '%стажировка%'
    or lower(v.name) like '%junior%')
   '''
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\2007790093.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

##### результат запроса
По полученным данным 51 вакансия подходит для начинающего дата-сайентиста, с учётом условий задания.
Также исходя из этого начинающему дата-сайентисту довольно сложно будет найти предстоящее место работы.

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [22]:
query_6_3 = f'''
    select
    count(*)
    from vacancies
    where (lower(name) like '%data science%'
    or lower(name) like '%data scientist%'
    or lower(name) like '%исследователь данных%'
    or (name) like '%ML%' and lower(name) not like '%html%'
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
    and (key_skills like '%SQL%' or lower(key_skills) like '%postgres%')
    '''
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\1100086283.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

##### результат запроса
C учётом критериев из предыдущего задания ищем вакансии для DS с ключевым навыком SQL или postgres.
Таких вакансий 201.

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [23]:
query_6_4 = f'''
    select
    count(*)
    from vacancies
    where (lower(name) like '%data science%'
    or lower(name) like '%data scientist%'
    or lower(name) like '%исследователь данных%'
    or (name) like '%ML%' and lower(name) not like '%html%'
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
    and key_skills like '%Python%'
        '''
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\156260367.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

##### результат запроса
В 351 вакансии у работодателей на профессию DS, в качестве ключевого навыка указан Python.

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [24]:
query_6_5 = f'''
    select
    avg(
    length(v.key_skills) - length(replace(v.key_skills, CHR(9), '')) +1
    )
    from public.vacancies v /*из таблицы вакансий*/
    where (lower(v.name) like '%data scientist%'
    or lower (v.name) like '%data science%'
    or lower (v.name) like '%исследователь данных%'
    or (v.name like '%ML%' and v.name not like '%HTML%')
    or lower (v.name) like '%machine learning%'
    or lower (v.name) like '%машинн%обучен%%')
    and v.key_skills is not NULL
    '''
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\4110880472.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg
0  6.406032

##### результат запроса
6.41 - это среднее количество ключевых навыков для DS, которые указывают работодатели в вакансиях.
Но возьмём целые числа, от 6 до 7, для удобства просмотра требований к ключевым навыкам.

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [25]:
query_6_6 = f'''
SELECT count(*), experience, ROUND(AVG((coalesce(salary_from, salary_to)+coalesce(salary_to, salary_from))/2))
    FROM vacancies
    where (lower(name) like '%data science%'
    or lower(name) like '%data scientist%'
    or lower(name) like '%исследователь данных%'
    or (name) like '%ML%' and lower(name) not like '%html%'
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%')
    AND ((salary_from IS NOT NULL) OR (salary_to IS NOT NULL))
    GROUP BY experience
    '''
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\ievge\AppData\Local\Temp\ipykernel_5856\2330117827.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


count          experience     round
0      7           Нет опыта   74643.0
1     28  От 1 года до 3 лет  139675.0
2     27       От 3 до 6 лет  243115.0

##### результат запроса
В коде выше получили данные из базы, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта, с учётом критериев необходимых при выполнении задания.
Для DS без опыта работы средняя заработная плата составляет 74643 Российских рубля, с опытом работы от 1 года до 3 лет - 139675 Российских рублей, с опытом работы от 3 до 6 лет - 243115 Российских рублей.

***

##### выводы по предметному анализу

Около 10% вакансий относятся к Junior и соискателям на профессию DS без опыта работы достаточно сложно найти работу.
Большинство вакансий требуют знания ключевых навыков, таких как Python, SQL, Postgres, также соискателям которые владеют этими навыками легче найти работу.
Также довольно сильный рост зарабртной платы при наличии опыта работы.

# Общий вывод по проекту

##### подведем итог исследования, обобщите выводы
##### здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

В итоге проекта мы выяснили следующее: количество вакансий в базе - 49197.

23501 - количество работодателей.

1362 - количество регионов, но не все они Российские, есть вакансии за рубежом, для более детального разбора требуется продолжения исследования данных.

Большинство вакансий - 35367 ищут DS на полный рабочий день и полную занятость.

Большинство вакансий c требованиями опыта работы от 1 года и выше.

Наибольшее количество предлагаемых ваканский относится к крупным городам.

Наибольшее количество вакансий среди компаний относятся к компании Яндекс.

Россия указывается в качестве некоторого района, что тоже требует более детального анализа.

8419 работодателей не указали сферу деятельности, что создаёт небольшую проблему при выборе.

Некий Поляков Денис Иванович разместил вакансии в 88 регионах, на мой взгляд это похоже на индивидуального предпринимателя, но слишком странно для ИП иметь настолько обширную сеть, подозрения что это выброс, но требуется более подробно раобрать эту структуру.

Также требуется примерно 6-7 навыков для работы DS, и тут тоже нужно дополнительные исследования для того, чтобы разобрать с каким сферам деятельности и с какими условиями это сочетается.

In [26]:
connection.close()
